language-plutus-core/src/Language/PlutusCore/Normalize.hs

```diff
@@ -0,0 +1,72 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE TypeFamilies          #-}
+
+-- | This module makes sure terms and types are well-formed according to Fig. 2
+module Language.PlutusCore.Normalize ( normalize
+                                     , NormalizationError (..)
+                                     ) where
+
+import           Data.Functor.Foldable
+import           Data.Functor.Foldable.Monadic
+import           Language.PlutusCore.Type
+import           PlutusPrelude
+
+data NormalizationError a = BadType a
+                          | BadTerm a
+
+instance Pretty a => Pretty (NormalizationError a) where
+    pretty (BadType l) = "Malformed type at" <+> pretty l
+    pretty (BadTerm l) = "Malformed term at" <+> pretty l
+
+-- | Ensure that all terms and types are well-formed accoring to Fig. 2
+normalize :: Program tyname name a -> Either (NormalizationError a) (Program tyname name a)
```

Comment:
I am not sure what the intent here is. The name and the type of the function suggest that it is performing `normalisation`, which is not what we want. We just want to *check* that a program is well-formed. This means that the result can just be a `Maybe Error`.